File: frontend/src/lib/investigation-pipeline.ts
import type {
  InvestigationPipelineUpdate,
  SankeyLinkConfig,
  SankeyNodeConfig,
} from "@/lib/types";
import {
  INVESTIGATION_METRICS,
  INVESTIGATION_SANKEY_LINKS,
  INVESTIGATION_SANKEY_NODES,
} from "@/data/investigation-sankey";

export function getDefaultPipelineUpdate(): InvestigationPipelineUpdate {
  return {
    nodes: INVESTIGATION_SANKEY_NODES.map((n) => ({ ...n })),
    links: INVESTIGATION_SANKEY_LINKS.map((l) => ({ ...l })),
    metrics: { ...INVESTIGATION_METRICS },
  };
}

export function validatePipelineConfig(body: InvestigationPipelineUpdate): string[] {
  const errors: string[] = [];
  if (!body.nodes.length) errors.push("At least one node is required");
  const ids = body.nodes.map((n) => n.id.trim());
  if (ids.some((id) => !id)) errors.push("Node id cannot be empty");
  if (new Set(ids).size !== ids.length) errors.push("Node ids must be unique");
  const idSet = new Set(ids);
  if (!body.links.length) errors.push("At least one link is required");
  body.links.forEach((link, i) => {
    if (!idSet.has(link.source)) {
      errors.push(`Link #${i + 1}: source "${link.source}" not found in nodes`);
    }
    if (!idSet.has(link.target)) {
      errors.push(`Link #${i + 1}: target "${link.target}" not found in nodes`);
    }
    if (link.value < 0) errors.push(`Link #${i + 1}: value must be >= 0`);
  });
  body.nodes.forEach((node, i) => {
    if (node.value < 0) errors.push(`Node #${i + 1}: value must be >= 0`);
    if (node.column < 0) errors.push(`Node #${i + 1}: column must be >= 0`);
    if (node.row < 0) errors.push(`Node #${i + 1}: row must be >= 0`);
  });
  return errors;
}

export function newEmptyNode(column = 0, row = 0): SankeyNodeConfig {
  const id = `node-${Date.now()}`;
  return { id, label: "New node", value: 10, column, row, color: "#6B5CE7" };
}

export function newEmptyLink(source = "", target = ""): SankeyLinkConfig {
  return { source, target, value: 5, color: "#6B5CE7" };
}

File: backend/app/services/investigation_pipeline.py
"""Investigation pipeline config — defaults, validation, SQLite persistence."""
from __future__ import annotations

from fastapi import HTTPException

from .. import db
from ..data.investigation_pipeline_defaults import (
    DEFAULT_LINKS,
    DEFAULT_METRICS,
    DEFAULT_NODES,
)
from ..models import (
    InvestigationPipelineConfig,
    InvestigationPipelineUpdate,
)


def _defaults_response() -> InvestigationPipelineConfig:
    return InvestigationPipelineConfig(
        configured=False,
        nodes=DEFAULT_NODES,
        links=DEFAULT_LINKS,
        metrics=DEFAULT_METRICS,
    )


def validate_pipeline(body: InvestigationPipelineUpdate) -> list[str]:
    errors: list[str] = []
    if not body.nodes:
        errors.append("At least one node is required")
    ids = [n.id.strip() for n in body.nodes]
    if any(not i for i in ids):
        errors.append("Node id cannot be empty")
    if len(ids) != len(set(ids)):
        errors.append("Node ids must be unique")
    id_set = set(ids)
    if not body.links:
        errors.append("At least one link is required")
    for i, link in enumerate(body.links):
        if link.source not in id_set:
            errors.append(f"Link #{i + 1}: source '{link.source}' not found in nodes")
        if link.target not in id_set:
            errors.append(f"Link #{i + 1}: target '{link.target}' not found in nodes")
        if link.value < 0:
            errors.append(f"Link #{i + 1}: value must be >= 0")
    for i, node in enumerate(body.nodes):
        if node.value < 0:
            errors.append(f"Node #{i + 1}: value must be >= 0")
        if node.column < 0:
            errors.append(f"Node #{i + 1}: column must be >= 0")
        if node.row < 0:
            errors.append(f"Node #{i + 1}: row must be >= 0")
    return errors


def get_pipeline_config() -> InvestigationPipelineConfig:
    raw = db.get_investigation_pipeline()
    if not raw:
        return _defaults_response()
    return InvestigationPipelineConfig(
        configured=True,
        nodes=raw.get("nodes", DEFAULT_NODES),
        links=raw.get("links", DEFAULT_LINKS),
        metrics=raw.get("metrics"),
    )


def save_pipeline_config(
    body: InvestigationPipelineUpdate,
) -> InvestigationPipelineConfig:
    errors = validate_pipeline(body)
    if errors:
        raise HTTPException(status_code=422, detail="; ".join(errors))
    payload = {
        "nodes": [n.model_dump() for n in body.nodes],
        "links": [l.model_dump() for l in body.links],
        "metrics": body.metrics.model_dump() if body.metrics else None,
    }
    db.set_investigation_pipeline(payload)
    return InvestigationPipelineConfig(configured=True, **payload)


def reset_pipeline_config() -> InvestigationPipelineConfig:
    db.clear_investigation_pipeline()
    return _defaults_response()
